psa-crypto/src/operations/key_management.rs

```diff
@@ -227,17 +219,87 @@ pub fn export_public(key: Id, data: &mut [u8]) -> Result<usize> {
     let handle = key.handle()?;
     let mut data_length = 0;
 
-    Status::from(unsafe {
+    let export_res = Status::from(unsafe {
         psa_crypto_sys::psa_export_public_key(
             handle,
             data.as_mut_ptr(),
             data.len(),
             &mut data_length,
         )
     })
-    .to_result()?;
+    .to_result();
+    key.close_handle(handle)?;
+    export_res?;
+    Ok(data_length)
+}
 
+/// Gets the attributes for a given key ID
+///
+/// The `Id` structure can be created with the `from_persistent_key_id` constructor on `Id`.
+///
+/// # Example
+///
+/// ```
+/// # use psa_crypto::operations::key_management;
+/// # use psa_crypto::types::key::{Attributes, Type, Lifetime, Policy, UsageFlags};
+/// # use psa_crypto::types::algorithm::{AsymmetricSignature, Hash};
+/// # let mut attributes = Attributes {
+/// #     key_type: Type::RsaKeyPair,
+/// #     bits: 1024,
+/// #     lifetime: Lifetime::Volatile,
+/// #     policy: Policy {
+/// #         usage_flags: UsageFlags {
+/// #             sign_hash: true,
+/// #             sign_message: true,
+/// #             verify_hash: true,
+/// #             verify_message: true,
+/// #             ..Default::default()
+/// #         },
+/// #         permitted_algorithms: AsymmetricSignature::RsaPkcs1v15Sign {
+/// #             hash_alg: Hash::Sha256.into(),
+/// #         }.into(),
+/// #     },
+/// # };
+/// psa_crypto::init().unwrap();
+/// let my_key = key_management::generate(attributes, None).unwrap();
+/// //...
+/// let key_attributes = key_management::get_key_attributes(my_key);
+/// ```
+
+pub fn get_key_attributes(key: Id) -> Result<Attributes> {
+    initialized()?;
+    let mut key_attributes = unsafe { psa_crypto_sys::psa_key_attributes_init() };
```

Comment:
I think if we follow the specs, the attributes created here might allocate some extra space that needs to be freed with a call to `psa_reset_key_attributes` (`Attributes::reset`). I think that might be needed on every path leaving the function once `psa_key_attributes_init` has been called.

Reply:
Hence my pet peeve with these cleanup methods.
What happens if both `close_handle` and `psa_reset_key_attributes` fail, which value do we return? Maybe we should really return a generic error when some of these "internal" methods fail, that the user should not really know about. 

An alternative would be to make a "layered" function, where for each variable that needs cleaning up we wrap everything after it in a function that returns a `Result`. If the result is `Ok`, attempt to clean up and if that fails return an error, otherwise return the `Ok` you got. If what you got was `Err`, then do the clean up anyway and return the `Err` you got from the next layer function. If you have two things that you need to clean, unfortunately you'd have 2 "layer" functions. But the benefit is that you're sure that you clean up everything and you return the `Err` that's most appropriate.

Reply:
I think `psa_reset_key_attributes` can't fail, though that doesn't invalidate the general point, just that particular example!

Reply:
(Also, I think you don't need to call `psa_reset_key_attributes` if calling `psa_key_attributes_init` is all you've done with the `psa_key_attributes_t`, and according to the 1.0 spec there seem to be certain `psa_set_key_` functions that you can call without needing to do the reset.)